// angular2/src/reflection/reflection_capabilities.js
var lang_1 = require('angular2/src/facade/lang');
var collection_1 = require('angular2/src/facade/collection');
// HACK: workaround for Traceur behavior.
// It expects all transpiled modules to contain this marker.
// TODO: remove this when we no longer use traceur
exports.__esModule = true;
var ReflectionCapabilities = (function () {
    function ReflectionCapabilities() {
    }
    ReflectionCapabilities.prototype.factory = function (type) {
        switch (type.length) {
            case 0:
                return function () {
                    return new type();
                };
            case 1:
                return function (a1) {
                    return new type(a1);
                };
            case 2:
                return function (a1, a2) {
                    return new type(a1, a2);
                };
            case 3:
                return function (a1, a2, a3) {
                    return new type(a1, a2, a3);
                };
            case 4:
                return function (a1, a2, a3, a4) {
                    return new type(a1, a2, a3, a4);
                };
            case 5:
                return function (a1, a2, a3, a4, a5) {
                    return new type(a1, a2, a3, a4, a5);
                };
            case 6:
                return function (a1, a2, a3, a4, a5, a6) {
                    return new type(a1, a2, a3, a4, a5, a6);
                };
            case 7:
                return function (a1, a2, a3, a4, a5, a6, a7) {
                    return new type(a1, a2, a3, a4, a5, a6, a7);
                };
            case 8:
                return function (a1, a2, a3, a4, a5, a6, a7, a8) {
                    return new type(a1, a2, a3, a4, a5, a6, a7, a8);
                };
            case 9:
                return function (a1, a2, a3, a4, a5, a6, a7, a8, a9) {
                    return new type(a1, a2, a3, a4, a5, a6, a7, a8, a9);
                };
            case 10:
                return function (a1, a2, a3, a4, a5, a6, a7, a8, a9, a10) {
                    return new type(a1, a2, a3, a4, a5, a6, a7, a8, a9, a10);
                };
        }
        ;
        throw new Error("Factory cannot take more than 10 arguments");
    };
    ReflectionCapabilities.prototype.parameters = function (typeOfFunc) {
        return lang_1.isPresent(typeOfFunc.parameters) ? typeOfFunc.parameters : collection_1.ListWrapper.createFixedSize(typeOfFunc.length);
    };
    ReflectionCapabilities.prototype.annotations = function (typeOfFunc) {
        return lang_1.isPresent(typeOfFunc.annotations) ? typeOfFunc.annotations : [];
    };
    ReflectionCapabilities.prototype.getter = function (name) {
        return new Function('o', 'return o.' + name + ';');
    };
    ReflectionCapabilities.prototype.setter = function (name) {
        return new Function('o', 'v', 'return o.' + name + ' = v;');
    };
    ReflectionCapabilities.prototype.method = function (name) {
        var functionBody = "if (!o." + name + ") throw new Error('\"" + name + "\" is undefined');\n        return o." + name + ".apply(o, args);";
        return new Function('o', 'args', functionBody);
    };
    return ReflectionCapabilities;
})();
exports.ReflectionCapabilities = ReflectionCapabilities;
