// angular2/src/test_lib/test_bed.js
"use strict";
Object.defineProperties(module.exports, {
  TestBed: {get: function() {
      return TestBed;
    }},
  ViewProxy: {get: function() {
      return ViewProxy;
    }},
  __esModule: {value: true}
});
var $__angular2_47_di__,
    $__angular2_47_src_47_facade_47_lang__,
    $__angular2_47_src_47_facade_47_async__,
    $__angular2_47_src_47_facade_47_lang__,
    $__angular2_47_src_47_facade_47_collection__,
    $__angular2_47_src_47_core_47_annotations_47_view__,
    $__angular2_47_src_47_core_47_compiler_47_template_95_resolver__,
    $__angular2_47_src_47_core_47_compiler_47_view__,
    $__angular2_47_src_47_core_47_compiler_47_dynamic_95_component_95_loader__,
    $__utils__,
    $__lang_95_utils__;
var $__0 = ($__angular2_47_di__ = require("angular2/di"), $__angular2_47_di__ && $__angular2_47_di__.__esModule && $__angular2_47_di__ || {default: $__angular2_47_di__}),
    Injector = $__0.Injector,
    bind = $__0.bind;
var $__1 = ($__angular2_47_src_47_facade_47_lang__ = require("angular2/src/facade/lang"), $__angular2_47_src_47_facade_47_lang__ && $__angular2_47_src_47_facade_47_lang__.__esModule && $__angular2_47_src_47_facade_47_lang__ || {default: $__angular2_47_src_47_facade_47_lang__}),
    Type = $__1.Type,
    isPresent = $__1.isPresent,
    BaseException = $__1.BaseException;
var Promise = ($__angular2_47_src_47_facade_47_async__ = require("angular2/src/facade/async"), $__angular2_47_src_47_facade_47_async__ && $__angular2_47_src_47_facade_47_async__.__esModule && $__angular2_47_src_47_facade_47_async__ || {default: $__angular2_47_src_47_facade_47_async__}).Promise;
var isBlank = ($__angular2_47_src_47_facade_47_lang__ = require("angular2/src/facade/lang"), $__angular2_47_src_47_facade_47_lang__ && $__angular2_47_src_47_facade_47_lang__.__esModule && $__angular2_47_src_47_facade_47_lang__ || {default: $__angular2_47_src_47_facade_47_lang__}).isBlank;
var List = ($__angular2_47_src_47_facade_47_collection__ = require("angular2/src/facade/collection"), $__angular2_47_src_47_facade_47_collection__ && $__angular2_47_src_47_facade_47_collection__.__esModule && $__angular2_47_src_47_facade_47_collection__ || {default: $__angular2_47_src_47_facade_47_collection__}).List;
var View = ($__angular2_47_src_47_core_47_annotations_47_view__ = require("angular2/src/core/annotations/view"), $__angular2_47_src_47_core_47_annotations_47_view__ && $__angular2_47_src_47_core_47_annotations_47_view__.__esModule && $__angular2_47_src_47_core_47_annotations_47_view__ || {default: $__angular2_47_src_47_core_47_annotations_47_view__}).View;
var TemplateResolver = ($__angular2_47_src_47_core_47_compiler_47_template_95_resolver__ = require("angular2/src/core/compiler/template_resolver"), $__angular2_47_src_47_core_47_compiler_47_template_95_resolver__ && $__angular2_47_src_47_core_47_compiler_47_template_95_resolver__.__esModule && $__angular2_47_src_47_core_47_compiler_47_template_95_resolver__ || {default: $__angular2_47_src_47_core_47_compiler_47_template_95_resolver__}).TemplateResolver;
var AppView = ($__angular2_47_src_47_core_47_compiler_47_view__ = require("angular2/src/core/compiler/view"), $__angular2_47_src_47_core_47_compiler_47_view__ && $__angular2_47_src_47_core_47_compiler_47_view__.__esModule && $__angular2_47_src_47_core_47_compiler_47_view__ || {default: $__angular2_47_src_47_core_47_compiler_47_view__}).AppView;
var $__8 = ($__angular2_47_src_47_core_47_compiler_47_dynamic_95_component_95_loader__ = require("angular2/src/core/compiler/dynamic_component_loader"), $__angular2_47_src_47_core_47_compiler_47_dynamic_95_component_95_loader__ && $__angular2_47_src_47_core_47_compiler_47_dynamic_95_component_95_loader__.__esModule && $__angular2_47_src_47_core_47_compiler_47_dynamic_95_component_95_loader__ || {default: $__angular2_47_src_47_core_47_compiler_47_dynamic_95_component_95_loader__}),
    DynamicComponentLoader = $__8.DynamicComponentLoader,
    ComponentRef = $__8.ComponentRef;
var $__9 = ($__utils__ = require("./utils"), $__utils__ && $__utils__.__esModule && $__utils__ || {default: $__utils__}),
    queryView = $__9.queryView,
    viewRootNodes = $__9.viewRootNodes,
    el = $__9.el;
var $__10 = ($__lang_95_utils__ = require("./lang_utils"), $__lang_95_utils__ && $__lang_95_utils__.__esModule && $__lang_95_utils__ || {default: $__lang_95_utils__}),
    instantiateType = $__10.instantiateType,
    getTypeOf = $__10.getTypeOf;
var TestBed = function TestBed(injector) {
  this._injector = injector;
};
($traceurRuntime.createClass)(TestBed, {
  overrideView: function(component, template) {
    this._injector.get(TemplateResolver).setView(component, template);
  },
  setInlineTemplate: function(component, html) {
    this._injector.get(TemplateResolver).setInlineTemplate(component, html);
  },
  overrideDirective: function(component, from, to) {
    this._injector.get(TemplateResolver).overrideTemplateDirective(component, from, to);
  },
  createView: function(component) {
    var $__13,
        $__14;
    var $__12 = arguments[1] !== (void 0) ? arguments[1] : {},
        context = ($__13 = $__12.context) === void 0 ? null : $__13,
        html = ($__14 = $__12.html) === void 0 ? null : $__14;
    if (isBlank(component) && isBlank(context)) {
      throw new BaseException('You must specified at least a component or a context');
    }
    if (isBlank(component)) {
      component = getTypeOf(context);
    } else if (isBlank(context)) {
      context = instantiateType(component);
    }
    if (isPresent(html)) {
      this.setInlineTemplate(component, html);
    }
    var rootEl = el('<div></div>');
    var componentBinding = bind(component).toValue(context);
    return this._injector.get(DynamicComponentLoader).loadIntoNewLocation(componentBinding, null, rootEl, this._injector).then((function(hostComponentRef) {
      return new ViewProxy(hostComponentRef);
    }));
  }
}, {});
Object.defineProperty(TestBed, "parameters", {get: function() {
    return [[Injector]];
  }});
Object.defineProperty(TestBed.prototype.overrideView, "parameters", {get: function() {
    return [[Type], [View]];
  }});
Object.defineProperty(TestBed.prototype.setInlineTemplate, "parameters", {get: function() {
    return [[Type], [$traceurRuntime.type.string]];
  }});
Object.defineProperty(TestBed.prototype.overrideDirective, "parameters", {get: function() {
    return [[Type], [Type], [Type]];
  }});
Object.defineProperty(TestBed.prototype.createView, "parameters", {get: function() {
    return [[Type], []];
  }});
var ViewProxy = function ViewProxy(componentRef) {
  this._componentRef = componentRef;
  this._view = componentRef.hostView.componentChildViews[0];
};
($traceurRuntime.createClass)(ViewProxy, {
  get context() {
    return this._view.context;
  },
  get rootNodes() {
    return viewRootNodes(this._view);
  },
  detectChanges: function() {
    this._view.changeDetector.detectChanges();
  },
  querySelector: function(selector) {
    return queryView(this._view, selector);
  },
  destroy: function() {
    this._componentRef.dispose();
  },
  get rawView() {
    return this._view;
  }
}, {});
Object.defineProperty(ViewProxy, "parameters", {get: function() {
    return [[ComponentRef]];
  }});
