// angular2/src/core/compiler/dynamic_component_loader.js
"use strict";
Object.defineProperties(module.exports, {
  ComponentRef: {get: function() {
      return ComponentRef;
    }},
  DynamicComponentLoader: {get: function() {
      return DynamicComponentLoader;
    }},
  __esModule: {value: true}
});
var $__angular2_47_di__,
    $__compiler__,
    $__angular2_47_src_47_facade_47_lang__,
    $__angular2_47_src_47_facade_47_async__,
    $__angular2_47_src_47_core_47_compiler_47_view_95_manager__,
    $__element_95_injector__,
    $__view__;
var $__0 = ($__angular2_47_di__ = require("angular2/di"), $__angular2_47_di__ && $__angular2_47_di__.__esModule && $__angular2_47_di__ || {default: $__angular2_47_di__}),
    Key = $__0.Key,
    Injector = $__0.Injector,
    Injectable = $__0.Injectable,
    ResolvedBinding = $__0.ResolvedBinding,
    Binding = $__0.Binding,
    bind = $__0.bind;
var Compiler = ($__compiler__ = require("./compiler"), $__compiler__ && $__compiler__.__esModule && $__compiler__ || {default: $__compiler__}).Compiler;
var $__2 = ($__angular2_47_src_47_facade_47_lang__ = require("angular2/src/facade/lang"), $__angular2_47_src_47_facade_47_lang__ && $__angular2_47_src_47_facade_47_lang__.__esModule && $__angular2_47_src_47_facade_47_lang__ || {default: $__angular2_47_src_47_facade_47_lang__}),
    Type = $__2.Type,
    BaseException = $__2.BaseException,
    stringify = $__2.stringify,
    isPresent = $__2.isPresent;
var Promise = ($__angular2_47_src_47_facade_47_async__ = require("angular2/src/facade/async"), $__angular2_47_src_47_facade_47_async__ && $__angular2_47_src_47_facade_47_async__.__esModule && $__angular2_47_src_47_facade_47_async__ || {default: $__angular2_47_src_47_facade_47_async__}).Promise;
var AppViewManager = ($__angular2_47_src_47_core_47_compiler_47_view_95_manager__ = require("angular2/src/core/compiler/view_manager"), $__angular2_47_src_47_core_47_compiler_47_view_95_manager__ && $__angular2_47_src_47_core_47_compiler_47_view_95_manager__.__esModule && $__angular2_47_src_47_core_47_compiler_47_view_95_manager__ || {default: $__angular2_47_src_47_core_47_compiler_47_view_95_manager__}).AppViewManager;
var ElementRef = ($__element_95_injector__ = require("./element_injector"), $__element_95_injector__ && $__element_95_injector__.__esModule && $__element_95_injector__ || {default: $__element_95_injector__}).ElementRef;
var AppView = ($__view__ = require("./view"), $__view__ && $__view__.__esModule && $__view__ || {default: $__view__}).AppView;
var ComponentRef = function ComponentRef(location, instance, componentView, dispose) {
  this.location = location;
  this.instance = instance;
  this.componentView = componentView;
  this._dispose = dispose;
};
($traceurRuntime.createClass)(ComponentRef, {
  get injector() {
    return this.location.injector;
  },
  get hostView() {
    return this.location.hostView;
  },
  dispose: function() {
    this._dispose();
  }
}, {});
Object.defineProperty(ComponentRef, "parameters", {get: function() {
    return [[ElementRef], [$traceurRuntime.type.any], [AppView], [Function]];
  }});
var DynamicComponentLoader = function DynamicComponentLoader(compiler, viewManager) {
  this._compiler = compiler;
  this._viewManager = viewManager;
};
($traceurRuntime.createClass)(DynamicComponentLoader, {
  loadIntoExistingLocation: function(typeOrBinding, location) {
    var injector = arguments[2] !== (void 0) ? arguments[2] : null;
    var $__7 = this;
    var binding = this._getBinding(typeOrBinding);
    return this._compiler.compile(binding.token).then((function(componentProtoView) {
      var componentView = $__7._viewManager.createDynamicComponentView(location, componentProtoView, binding, injector);
      var dispose = (function() {
        throw new BaseException("Not implemented");
      });
      return new ComponentRef(location, location.elementInjector.getDynamicallyLoadedComponent(), componentView, dispose);
    }));
  },
  loadIntoNewLocation: function(typeOrBinding, parentComponentLocation, elementOrSelector) {
    var injector = arguments[3] !== (void 0) ? arguments[3] : null;
    var $__7 = this;
    return this._compiler.compileInHost(this._getBinding(typeOrBinding)).then((function(hostProtoView) {
      var hostView = $__7._viewManager.createInPlaceHostView(parentComponentLocation, elementOrSelector, hostProtoView, injector);
      var newLocation = hostView.elementInjectors[0].getElementRef();
      var component = hostView.elementInjectors[0].getComponent();
      var dispose = (function() {
        $__7._viewManager.destroyInPlaceHostView(parentComponentLocation, hostView);
      });
      return new ComponentRef(newLocation, component, hostView.componentChildViews[0], dispose);
    }));
  },
  loadNextToExistingLocation: function(typeOrBinding, location) {
    var injector = arguments[2] !== (void 0) ? arguments[2] : null;
    var binding = this._getBinding(typeOrBinding);
    return this._compiler.compileInHost(binding).then((function(hostProtoView) {
      var hostView = location.viewContainer.create(-1, hostProtoView, injector);
      var newLocation = hostView.elementInjectors[0].getElementRef();
      var component = hostView.elementInjectors[0].getComponent();
      var dispose = (function() {
        var index = location.viewContainer.indexOf(hostView);
        location.viewContainer.remove(index);
      });
      return new ComponentRef(newLocation, component, hostView.componentChildViews[0], dispose);
    }));
  },
  _getBinding: function(typeOrBinding) {
    var binding;
    if (typeOrBinding instanceof Binding) {
      binding = typeOrBinding;
    } else {
      binding = bind(typeOrBinding).toClass(typeOrBinding);
    }
    return binding;
  }
}, {});
Object.defineProperty(DynamicComponentLoader, "annotations", {get: function() {
    return [new Injectable()];
  }});
Object.defineProperty(DynamicComponentLoader, "parameters", {get: function() {
    return [[Compiler], [AppViewManager]];
  }});
Object.defineProperty(DynamicComponentLoader.prototype.loadIntoExistingLocation, "parameters", {get: function() {
    return [[], [ElementRef], [Injector]];
  }});
Object.defineProperty(DynamicComponentLoader.prototype.loadIntoNewLocation, "parameters", {get: function() {
    return [[], [ElementRef], [$traceurRuntime.type.any], [Injector]];
  }});
Object.defineProperty(DynamicComponentLoader.prototype.loadNextToExistingLocation, "parameters", {get: function() {
    return [[], [ElementRef], [Injector]];
  }});
