// angular2/src/core/compiler/view_container_ref.js
"use strict";
Object.defineProperties(module.exports, {
  ViewContainerRef: {get: function() {
      return ViewContainerRef;
    }},
  __esModule: {value: true}
});
var $__angular2_47_src_47_facade_47_collection__,
    $__angular2_47_di__,
    $__angular2_47_src_47_core_47_compiler_47_element_95_injector__,
    $__angular2_47_src_47_facade_47_lang__,
    $__view__,
    $__view_95_manager__;
var $__0 = ($__angular2_47_src_47_facade_47_collection__ = require("angular2/src/facade/collection"), $__angular2_47_src_47_facade_47_collection__ && $__angular2_47_src_47_facade_47_collection__.__esModule && $__angular2_47_src_47_facade_47_collection__ || {default: $__angular2_47_src_47_facade_47_collection__}),
    ListWrapper = $__0.ListWrapper,
    MapWrapper = $__0.MapWrapper,
    List = $__0.List;
var Injector = ($__angular2_47_di__ = require("angular2/di"), $__angular2_47_di__ && $__angular2_47_di__.__esModule && $__angular2_47_di__ || {default: $__angular2_47_di__}).Injector;
var eiModule = ($__angular2_47_src_47_core_47_compiler_47_element_95_injector__ = require("angular2/src/core/compiler/element_injector"), $__angular2_47_src_47_core_47_compiler_47_element_95_injector__ && $__angular2_47_src_47_core_47_compiler_47_element_95_injector__.__esModule && $__angular2_47_src_47_core_47_compiler_47_element_95_injector__ || {default: $__angular2_47_src_47_core_47_compiler_47_element_95_injector__});
var $__2 = ($__angular2_47_src_47_facade_47_lang__ = require("angular2/src/facade/lang"), $__angular2_47_src_47_facade_47_lang__ && $__angular2_47_src_47_facade_47_lang__.__esModule && $__angular2_47_src_47_facade_47_lang__ || {default: $__angular2_47_src_47_facade_47_lang__}),
    isPresent = $__2.isPresent,
    isBlank = $__2.isBlank;
var viewModule = ($__view__ = require("./view"), $__view__ && $__view__.__esModule && $__view__ || {default: $__view__});
var avmModule = ($__view_95_manager__ = require("./view_manager"), $__view_95_manager__ && $__view_95_manager__.__esModule && $__view_95_manager__ || {default: $__view_95_manager__});
var ViewContainerRef = function ViewContainerRef(viewManager, location, defaultProtoView) {
  this._viewManager = viewManager;
  this._location = location;
  this._defaultProtoView = defaultProtoView;
};
($traceurRuntime.createClass)(ViewContainerRef, {
  _getViews: function() {
    var vc = this._location.hostView.viewContainers[this._location.boundElementIndex];
    return isPresent(vc) ? vc.views : [];
  },
  clear: function() {
    for (var i = this.length - 1; i >= 0; i--) {
      this.remove(i);
    }
  },
  get: function(index) {
    return this._getViews()[index];
  },
  get length() {
    return this._getViews().length;
  },
  create: function() {
    var atIndex = arguments[0] !== (void 0) ? arguments[0] : -1;
    var protoView = arguments[1] !== (void 0) ? arguments[1] : null;
    var injector = arguments[2] !== (void 0) ? arguments[2] : null;
    if (atIndex == -1)
      atIndex = this.length;
    if (isBlank(protoView)) {
      protoView = this._defaultProtoView;
    }
    return this._viewManager.createViewInContainer(this._location, atIndex, protoView, injector);
  },
  insert: function(view) {
    var atIndex = arguments[1] !== (void 0) ? arguments[1] : -1;
    if (atIndex == -1)
      atIndex = this.length;
    return this._viewManager.attachViewInContainer(this._location, atIndex, view);
  },
  indexOf: function(view) {
    return ListWrapper.indexOf(this._getViews(), view);
  },
  remove: function() {
    var atIndex = arguments[0] !== (void 0) ? arguments[0] : -1;
    if (atIndex == -1)
      atIndex = this.length - 1;
    this._viewManager.destroyViewInContainer(this._location, atIndex);
  },
  detach: function() {
    var atIndex = arguments[0] !== (void 0) ? arguments[0] : -1;
    if (atIndex == -1)
      atIndex = this.length - 1;
    return this._viewManager.detachViewInContainer(this._location, atIndex);
  }
}, {});
Object.defineProperty(ViewContainerRef, "parameters", {get: function() {
    return [[avmModule.AppViewManager], [eiModule.ElementRef], [viewModule.AppProtoView]];
  }});
Object.defineProperty(ViewContainerRef.prototype.get, "parameters", {get: function() {
    return [[$traceurRuntime.type.number]];
  }});
Object.defineProperty(ViewContainerRef.prototype.create, "parameters", {get: function() {
    return [[$traceurRuntime.type.number], [viewModule.AppProtoView], [Injector]];
  }});
Object.defineProperty(ViewContainerRef.prototype.insert, "parameters", {get: function() {
    return [[viewModule.AppView], [$traceurRuntime.type.number]];
  }});
Object.defineProperty(ViewContainerRef.prototype.indexOf, "parameters", {get: function() {
    return [[viewModule.AppView]];
  }});
Object.defineProperty(ViewContainerRef.prototype.remove, "parameters", {get: function() {
    return [[$traceurRuntime.type.number]];
  }});
Object.defineProperty(ViewContainerRef.prototype.detach, "parameters", {get: function() {
    return [[$traceurRuntime.type.number]];
  }});
