// angular2/src/router/router_link.js
"use strict";
Object.defineProperties(module.exports, {
  RouterLink: {get: function() {
      return RouterLink;
    }},
  __esModule: {value: true}
});
var $__angular2_47_annotations__,
    $__angular2_47_core__,
    $__angular2_47_src_47_facade_47_lang__,
    $__angular2_47_src_47_dom_47_dom_95_adapter__,
    $__router__;
var Decorator = ($__angular2_47_annotations__ = require("angular2/annotations"), $__angular2_47_annotations__ && $__angular2_47_annotations__.__esModule && $__angular2_47_annotations__ || {default: $__angular2_47_annotations__}).Decorator;
var NgElement = ($__angular2_47_core__ = require("angular2/core"), $__angular2_47_core__ && $__angular2_47_core__.__esModule && $__angular2_47_core__ || {default: $__angular2_47_core__}).NgElement;
var isPresent = ($__angular2_47_src_47_facade_47_lang__ = require("angular2/src/facade/lang"), $__angular2_47_src_47_facade_47_lang__ && $__angular2_47_src_47_facade_47_lang__.__esModule && $__angular2_47_src_47_facade_47_lang__ || {default: $__angular2_47_src_47_facade_47_lang__}).isPresent;
var DOM = ($__angular2_47_src_47_dom_47_dom_95_adapter__ = require("angular2/src/dom/dom_adapter"), $__angular2_47_src_47_dom_47_dom_95_adapter__ && $__angular2_47_src_47_dom_47_dom_95_adapter__.__esModule && $__angular2_47_src_47_dom_47_dom_95_adapter__ || {default: $__angular2_47_src_47_dom_47_dom_95_adapter__}).DOM;
var Router = ($__router__ = require("./router"), $__router__ && $__router__.__esModule && $__router__ || {default: $__router__}).Router;
var RouterLink = function RouterLink(ngEl, router) {
  this._domEl = ngEl.domElement;
  this._router = router;
};
($traceurRuntime.createClass)(RouterLink, {
  set route(changes) {
    this._route = changes;
    this.updateHref();
  },
  set params(changes) {
    this._params = changes;
    this.updateHref();
  },
  updateHref: function() {
    if (isPresent(this._route) && isPresent(this._params)) {
      var newHref = this._router.generate(this._route, this._params);
      DOM.setAttribute(this._domEl, 'href', newHref);
    }
  }
}, {});
Object.defineProperty(RouterLink, "annotations", {get: function() {
    return [new Decorator({
      selector: '[router-link]',
      properties: {
        'route': 'routerLink',
        'params': 'routerParams'
      }
    })];
  }});
Object.defineProperty(RouterLink, "parameters", {get: function() {
    return [[NgElement], [Router]];
  }});
